File: tests/data/program_analysis/array_func_loop/array-to-func_06.f
C   File: array-to-func_06.f
C   This test is an example of passing arrays to a function (subroutine)
C   then pass the array to another function under the nested loop.
C
C   The code was sampled from SIR-Gillespie.f

      subroutine update_mean_var(MeanS, VarS, k, n, runs)
      integer, parameter :: Tmax = 100
      integer k, n, runs
      double precision, dimension(0:Tmax) :: MeanS, VarS

      MeanS(k) = MeanS(k) + (n - MeanS(k))/(runs+1)
      VarS(k) = VarS(k) + runs/(runs+1) * (n-MeanS(k))*(n-MeanS(k))

      return
      end subroutine update_mean_var

      subroutine gillespie(MeanS, VarS)
      integer, parameter :: Tmax = 100
      integer runs, sample, i, j

      double precision, dimension(0:Tmax) :: MeanS, VarS
      double precision, dimension(0:Tmax) :: samples

      do i = 0, Tmax
        samples(i) = i
      end do

      do runs = 0, Tmax
        j = 0
        do while (j .le. Tmax)
          call update_mean_var(MeanS, VarS, j, sample, runs)
          j = j + 1
        end do
        sample = samples(runs)
      end do

      end subroutine gillespie

      program main
      implicit none

      integer, parameter :: Tmax = 100
      double precision, dimension(0:Tmax) :: MeanS, VarS

      call gillespie(MeanS, VarS)

      end program main

File: tests/data/program_analysis/array_func_loop/array-to-func_06.py
import sys
import os
from typing import List
import math
from delphi.translators.for2py.format import *
from delphi.translators.for2py.arrays import *
from delphi.translators.for2py.static_save import *
from delphi.translators.for2py.strings import *
from delphi.translators.for2py import intrinsics
from dataclasses import dataclass
from delphi.translators.for2py.types_ext import Float32
import delphi.translators.for2py.math_ext as math
from numbers import Real
from random import random


def update_mean_var(means: Array, vars: Array, k: List[int], n: List[int], runs: List[int]):
    tmax: List[int] = [100]
    means.set_((k[0]), Float32((means.get_((k[0])) + ((n[0] - means.get_((k[0]))) / (runs[0] + 1)))))
    vars.set_((k[0]), Float32((vars.get_((k[0])) + (((runs[0] / (runs[0] + 1)) * (n[0] - means.get_((k[0])))) * (n[0] - means.get_((k[0])))))))
    

def gillespie(means: Array, vars: Array):
    tmax: List[int] = [100]
    runs: List[int] = [None]
    sample: List[int] = [None]
    i: List[int] = [None]
    j: List[int] = [None]
    samples = Array(float, [(0, tmax[0])])
    for i[0] in range(0, tmax[0]+1):
        samples.set_((i[0]), Float32(i[0]))
    for runs[0] in range(0, tmax[0]+1):
        j[0] = 0
        while (j[0] <= tmax[0]):
            update_mean_var(means, vars, j, sample, runs)
            j[0] = (j[0] + 1)
        sample[0] = samples.get_((runs[0]))

def main():
    tmax: List[int] = [100]
    means = Array(float, [(0, tmax[0])])
    vars = Array(float, [(0, tmax[0])])
    gillespie(means, vars)

main()
